pykickstart/commands/lang.py

```diff
@@ -55,9 +55,8 @@ def _getParser(self):
             will continue in English though the running system will have the
             specified langauge by default.
 
-            The file ``/usr/share/system-config-language/locale-list`` provides a
-            list the valid language codes in the first column of each line and
-            is part of the system-config-languages package.""", version=FC3)
+            You can list the valid language codes by calling
+            ``localectl list-locales``.""", version=FC3)
```

Comment:
We don't support all of them. It should be  `localectl list-locales | grep _`. Also `locale -a | grep _` works.

Reply:
Heh, True I've missed those "special" locales listed there. I wonder if there is not a better way to get the list. I'll investigate that a little bit.

Reply:
I was already looking into this for the RHEL documentation, but I couldn't find a better way. We are limited by the langtable, but the langtable doesn't support these special cases and doesn't provide a list of supported locales, so there is nothing else we can do.

Reply:
I want to ask creator of the library we are using where and how he is taking data. Just in case he would have better way to do that.

Reply:
> We don't support all of them. It should be `localectl list-locales | grep _`. Also `locale -a | grep _` works.

"localectl  list-locales" seems to do exactly the same as "locale -a". If I install extra locales by installing extra glibc locale packages, both list the additionally installed locales. I guess "localectl list-locales" just calls "locale -a". 

Most of the locale names listed by "locale -a" which do not contain an _ are legacy aliases.
But a few of them are  real, useful locales. These ones:

C
C.utf8
POSIX
eo.UTF-8  <- this is Esperanto

Reply:
> We don't support all of them.

I wonder  what you mean by "support".

Supported by glibc? That would be the output of "locale -a" or "localectl list-locales", but it depends on which glibc subpackages are installed. 

glibc-all-langpacks-2.29-15.fc30.x86_64 contains all of them. 

But one can also install only a few packages like 

glibc-langpack-en-2.29-15.fc30.x86_64
glibc-langpack-es-2.29-15.fc30.x86_64
...

then "locale -a" will list far fewer locales. 

Or do we mean by "supported" the languages where we have translations for? Or maybe the languages where anaconda has translations for?

At the start of the installation, anaconda shows a language selection page. The list of languages shown there is far smaller than the list of locales supported by glibc. 

As far as  I know this list is generated from the list of language where anaconda has translations for.

Reply:
> I was already looking into this for the RHEL documentation, but I couldn't find a better way. We are limited by the langtable, but the langtable doesn't support these special cases and doesn't provide a list of supported locales, so there is nothing else we can do.

I think the legacy aliases like "czech", "deutsch", etc... should be ignored and not used anymore.

eo.utf8 is a useful locale though and it is listed in langtable:

```
$ python3 
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import langtable
>>> langtable.list_locales(languageId='eo')
['eo.UTF-8']
```

Reply:
> I was already looking into this for the RHEL documentation, but I couldn't find a better way. We are limited by the langtable, but the langtable doesn't support these special cases and doesn't provide a list of supported locales, so there is nothing else we can do.

langtable tries to be as complete as possible in storing information about languages and territories. So langtable even includes languages where we don’t even have a locale in glibc.
For example:

```
$ python3 
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import langtable
>>> langtable.list_locales(languageId='av')
[]
>>> langtable.language_name(languageId='av')
'авар мацӀ'
>>> langtable.language_name(languageId='av', languageIdQuery='en')
'Avaric'
>>> 
```

So glibc currently has no locale available for “Avaric”. One could still store translations for “Avaric”
in

/usr/share/locale/av/LC_MESSAGES/

and set LANGUAGE="av" to get gettext translations for such a language which doesn’t yet have a glibc locale.

Currently, langtable does not have a function to return a list of locales supported by glibc.

Should I add such a function? Would that be helpful?

Currently, the list_locales() function returns nothing if called with an empty languageId or with no parameters at all:

```
>>> langtable.list_locales(languageId='')
[]
>>> langtable.list_locales()
[]
>>>
```

I could change that to list all available locales, i.e. all locales supported by glibc (without third party extras, users can add locales to glibc using localedef). Would that be helpful? Should I add something like that?

Reply:
Wow thanks a lot for your input @mike-fabian!

The problem is that Anaconda use other checks than you have. We are using this regex to parse the input for langtable or other usage:

https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/localization.py#L38

So basically we won't accept anything what don't have `_`. Not sure if that is a good way to go.


About your proposed solution. It won't help us in the code, we want that for users to know what they should specify for the installation to work. So we might have think about this as a binary what user will call to get list of supported languages but I would more prefer to stay as close as possible to the running system.

What do you think?

Reply:
I still don’t understand completely what is needed. Users who are going to do an installation and want to do what  languages are supported? Don’t they see that in the lists displayed on the first page shown by anaconda right at the start of the installation?

In the running system, "locale -a" shows the locales which are currently installed. Which could be all supported locales  if the glibc-all-langpacks package is installed. Or it could be a subset of these if only some glibc-langpack-?? packages are installed.

Reply:
> I still don’t understand completely what is needed. Users who are going to do an installation and want to do what languages are supported? Don’t they see that in the lists displayed on the first page shown by anaconda right at the start of the installation?

The problem is that user are able to decide the language based on the kernel boot parameter `inst.lang` and `inst.keyboard`. However, to be able to do that, they have to know the code for their language. We are trying to find out the best way to give them the list of what they can use.

> In the running system, "locale -a" shows the locales which are currently installed. Which could be all supported locales if the glibc-all-langpacks package is installed. Or it could be a subset of these if only some glibc-langpack-?? packages are installed.

So that means that user have to install all the `glibc-langpack-*` to get the correct list of what is supported in Anaconda? Of course with the good filters of the list (mean the ``grep |``).

Reply:
@mike-fabian ping, are you fine with merging the ``localectl list-locales | grep _`` solution?

Reply:
Yes, I think that is OK, although you miss eo.UTF-8 and  C.UTF-8. These are currently the only useful UTF-8 locales which don’t  have an _ in their locale name.